json_rules.py

```diff
@@ -0,0 +1,370 @@
+        if quotes not in {2, 4}:
+            return True, "quote_rule"
+    
+    return False, ""
+
+
+def breaks_weight_rule(json_lines: List[str]) -> Tuple[bool, str]:
+    """
+    Weight should be an int, a float or an expression.
+    In the first 2 cases, this rule wouldn't be broken. However,
+    we don't have expressions yet, so anything such as BW + 3 
+    would break this rule.
+
+    In:
+        json_lines: The lines which make up the message
+    
+    Out:
+        True, "weight_rule" if the rule written above is broken.
+        False, "" otherwise
+    """
+    weight_line = ""
+    for line in json_lines[1:-1]:
+        if "weight" in line:
+            weight_line = line
+            break
+    weight_value = weight_line.split(':')
+    value = weight_value[1].replace(',',"").strip()
+    try:
+        float(value)
+    except ValueError:
+        # If we can't convert automatically 
+        # to float, it's probably supposed
+        # to be an expression
+        return True, "weight_rule"
+    
+    return False, ""
+
+
+def breaks_misc_extra_symbols_rule(json_lines: List[str]) -> Tuple[bool, str]:
+    """
+    The only allowable symbols should be : +, -, ", :, [a-zA-Z0-9], ., <, >, {
+    }, , , \n, _,  ,
+
+    In:
+        json_lines: The lines which make up the message
+    
+    Out:
+        True, "misc_rule" if the rule written above is broken.
+        False, "" otherwise
+
+    """
+    
+    text = recreate_json(json_lines)
+    for pattern in ALLOWED_SYMBOLS_RE:
+        if len(text) == 0:
+            break
+        text = re.sub(pattern, "", text)
+        
+    
+    if len(text) == 0:
+        return False, ""
+
+    return True, "misc_rule"
+
+
+RULES = [
+    breaks_comma_rule,
+    breaks_braces_rule,
+    breaks_colon_rule,
+    breaks_quote_rule,
+    breaks_weight_rule,
+    breaks_misc_extra_symbols_rule,
+]
+
+
+def check_rules(json_lines: List[str]) -> Set[str]:
+    """
+    Takes as input the JSON lines and a set of rules to check.
+    The rules are callable functions which return tuples of type 
+    bool, str. 
+
+    In:
+        json_lines: The lines which make up the message
+
+    Out:
+        A set of all broken rules for this message
+    """
+    if not message_is_readable(json_lines):
+        return {"unreadable"}
+
+    broken_rules = set()
+    for check in RULES:
+        broken, rule = check(json_lines)
+        if broken:
+            broken_rules.add(rule)
+    return broken_rules
+
+total_broken = 0
+unreadable = 0
+
+with open(FILE_NAME, 'r') as file:
+    content = json.loads(file.read())
+    dictionary = dict()
+    for message in content["messages"]:
+        json_message = message["text"]
+        json_lines : List[str] = split_json_message_into_lines(json_message)
+        broken_rules = check_rules(json_lines)
+        total_broken += len(broken_rules) > 0
+        unreadable += "unreadable" in broken_rules
+        dictionary[message["id"]] = broken_rules
```

Comment:
Almost 60% (3076/5290) of the values in the dictionary in the main function consists of empty sets. Seems inefficient.